src/main/resources/stubs/sync/waitgroup.gobra

```diff
@@ -0,0 +1,123 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) 2011-2020 ETH Zurich.
+
+package sync
+
+// Definition according to https://github.com/golang/go/blob/master/src/sync/waitgroup.go
+type WaitGroup struct {
+	// noCopy noCopy
+	state1 [3]uint32
+}
+
+pred (wg *WaitGroup) WaitGroupP()
+pred (wg *WaitGroup) WaitGroupStarted()
+pred (wg *WaitGroup) UnitDebt(p pred())
+pred (wg *WaitGroup) Token(t pred())
+
+// TODO: to be removed when channels are added to the language, use the new keyword PredTrue instead
+pred pred_true() {
+	true
+}
+
+ghost
+requires acc(g.WaitGroupP(), _)
+pure func (g *WaitGroup) WaitMode() bool
+
+ghost
+requires acc(g) && *g == WaitGroup{}
+ensures g.WaitGroupP() && !g.WaitMode()
+func (g *WaitGroup) Init()
+
+ghost
+requires g.WaitGroupP()
+ensures g.WaitGroupP() && !g.WaitMode()
+func (g *WaitGroup) UnsetWaitMode()
+
+// The arguments `dividend` and `divisor` are the two components of a permission dividend/divisor.
+// This is a workaround for the lack of support for the perm type at the moment
+ghost
+requires dividend > 0 && divisor > 0 // p > 0
+requires acc(g.WaitGroupP(), dividend/divisor)
+requires !g.WaitMode() && g.UnitDebt(P)
+ensures g.UnitDebt(P) && acc(g.WaitGroupStarted(), dividend/divisor)
+func (g *WaitGroup) Start(ghost dividend int, ghost divisor int, ghost P pred())
+
+ghost
+requires dividendP >= 0 && divisorP > 0 // p >= 0
+requires dividendQ > 0 && divisorQ > 0 // q > 0
+requires dividendP * divisorQ + dividendQ * divisorP == divisorP * divisorQ // p + q == 1
+requires acc(g.WaitGroupP(), dividendP/divisorP)
+requires acc(g.WaitGroupStarted(), dividendQ/divisorQ)
+ensures g.WaitGroupP() && g.WaitMode()
+func (g *WaitGroup) SetWaitMode(ghost dividendP int, ghost divisorP int, ghost dividendQ int, ghost divisorQ int)
+
+// Simplified version of the debt redistribution rule. The most general version cannot be written in Gobra
+// without support for magic wands
+ghost
+requires P() && g.UnitDebt(P)
+ensures g.UnitDebt(pred_true!<!>)
+func (g *WaitGroup) PayDebt(ghost P pred())
+
+// Simplified version of the debt redistribution rule, instantiated with P == { pred_true!<!> } and Q == { R }.
+// This is the only rule that generates a Token.
+ghost
+requires R() && g.UnitDebt(pred_true!<!>)
+ensures g.UnitDebt(R) && g.Token(R)
+func (g *WaitGroup) GenerateToken(ghost R pred())
+
+requires dividend >= 0 && divisor > 0 // p >= 0
+requires n > 0 && dividend > 0 ==> acc(g.WaitGroupP(), dividend/divisor) && !g.WaitMode()
+requires (n > 0 && dividend == 0) ==> g.UnitDebt(P)
+requires n < 0 ==> acc(g.UnitDebt(pred_true!<!>), -n/1)
```

Comment:
Why is `pred_true` used here instead of `P` like in the document? Wouldn't this require some precondition that `P == PredTrue` or doesn't this matter since parameter P is never used if n is negative?

Reply:
Nevermind, I've just seen that this contract corresponds to the contract of `Add` in Remark 8.  It might help to add a smaller comment to this contract as well as the contract for `PayDebt` that they follow Remark 8

Reply:
I opted not to make references to particular points of Martin's draft because it is not publicly available and, in the future, the structure of the draft may change, invalidating our comments